File: leetcode/794-valid-tic-tac-toe-state/main.go
package main

import "fmt"

/*
	1st approach: very stupid checking

	Time	O(1) since board size is fixed
	Space	O(1) since board size is fixed
	0 ms, faster than 100.00%
*/
func validTicTacToe(board []string) bool {
	nx := 0
	no := 0
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			if board[i][j] == 'X' {
				nx++
			}
			if board[i][j] == 'O' {
				no++
			}
		}
	}
	if no > nx {
		return false
	} else if nx > no+1 {
		return false
	}
	winCountX := 0
	winCountO := 0
	// check rows and columns
	rowCountX := []int{0, 0, 0}
	colCountX := []int{0, 0, 0}
	rowCountO := []int{0, 0, 0}
	colCountO := []int{0, 0, 0}
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if board[i][j] == 'X' {
				rowCountX[i]++
				if rowCountX[i] == 3 {
					winCountX++
				}
				colCountX[j]++
				if colCountX[j] == 3 {
					winCountX++
				}
			}
			if board[i][j] == 'O' {
				rowCountO[i]++
				if rowCountO[j] == 3 {
					winCountO++
				}
				colCountO[j]++
				if colCountO[j] == 3 {
					winCountO++
				}
			}
		}
	}
	diagCountX := 0
	diagCountO := 0
	for i := 0; i < 3; i++ {
		if board[i][i] == 'X' {
			diagCountX++
			if diagCountX == 3 {
				winCountX++
			}
		}
		if board[i][i] == 'O' {
			diagCountO++
			if diagCountO == 3 {
				winCountO++
			}
		}
	}
	antiDiagCountX := 0
	antiDiagCountO := 0
	for i := 2; i >= 0; i-- {
		if board[i][2-i] == 'X' {
			antiDiagCountX++
			if antiDiagCountX == 3 {
				winCountX++
			}
		}
		if board[i][2-i] == 'O' {
			antiDiagCountO++
			if antiDiagCountO == 3 {
				winCountO++
			}
		}
	}
	if winCountX == 0 && winCountO == 0 {
		return true
	} else if winCountX == 0 && winCountO > 0 {
		if no < nx {
			return false
		}
		return true
	} else if winCountX > 0 && winCountO == 0 {
		if no == nx {
			return false
		}
		return true
	}
	return false
}

func main() {
	a := []string{
		"O  ", "   ", "   ",
	}
	fmt.Println(validTicTacToe(a))

	a = []string{
		"XOX", " X ", "   ",
	}
	fmt.Println(validTicTacToe(a))

	a = []string{
		"XXX", "   ", "OOO",
	}
	fmt.Println(validTicTacToe(a))

	a = []string{
		"XOX", "O O", "XOX",
	}
	fmt.Println(validTicTacToe(a))

	a = []string{
		"XXX", "OOX", "OOX",
	}
	fmt.Println(validTicTacToe(a))

	a = []string{
		"XXX", "XOO", "OO ",
	}
	fmt.Println(validTicTacToe(a))

	a = []string{
		"XXO", "XOX", "OXO",
	}
	fmt.Println(validTicTacToe(a))
}

File: leetcode/794-valid-tic-tac-toe-state/main.py
"""
    2nd approach:
    - if the number of symbols are valid, there are only 2 senarios
        1. player 1 wins and the number of X == O+1
        2. player 2 wins and the number of O == X

    Time    O(n^3) -> O(1) due to n=3 as always
    Space   O(2n) -> O(1) due to n=3 as always
    20 ms, faster than 82.47%
"""


class Solution(object):
    def validTicTacToe(self, board):
        """
        :type board: List[str]
        :rtype: bool
        """
        n = len(board)

        nx = 0
        no = 0

        rows = n*[0]
        cols = n*[0]
        diag = 0
        antiDiag = 0

        for i in range(len(board)):
            for j in range(len(board[i])):
                cell = board[i][j]
                toAdd = 0
                if cell == 'X':
                    nx += 1
                    toAdd = 1
                if cell == 'O':
                    no += 1
                    toAdd = -1
                rows[i] += toAdd
                cols[j] += toAdd
                if i == j:
                    diag += toAdd
                if i + j == n-1:
                    antiDiag += toAdd

        # if len no valid, return false immediately
        if nx > no + 1 or no > nx:
            return False

        player1win = False
        player2win = False

        for row in rows:
            if row == n:
                player1win = True
            if row == -n:
                player2win = True

        for col in cols:
            if col == n:
                player1win = True
            if col == -n:
                player2win = True

        if diag == n:
            player1win = True
        elif diag == -n:
            player2win = True

        if antiDiag == n:
            player1win = True
        elif antiDiag == -n:
            player2win = True

        if player1win and player2win:
            return False

        if player1win:
            if nx != no + 1:
                return False

        if player2win:
            if nx != no:
                return False

        return True


a = ["O  ", "   ", "   "]
print(Solution().validTicTacToe(a))

a = ["XOX", " X ", "   "]
print(Solution().validTicTacToe(a))

a = ["XXX", "   ", "OOO"]
print(Solution().validTicTacToe(a))

a = ["XOX", "O O", "XOX"]
print(Solution().validTicTacToe(a))

a = ["XXX", "OOX", "OOX"]
print(Solution().validTicTacToe(a))

a = ["XXX", "XOO", "OO "]
print(Solution().validTicTacToe(a))

a = ["XXO", "XOX", "OXO"]
print(Solution().validTicTacToe(a))
